src/ray/raylet/worker_pool.cc

```diff
@@ -198,30 +207,50 @@ pid_t WorkerPool::StartProcess(const std::vector<std::string> &worker_command_ar
     RAY_LOG(DEBUG) << stream.str();
   }
 
-  // Launch the process to create the worker.
-  pid_t pid = fork();
-
-  if (pid != 0) {
-    return pid;
+  pid_t pid;
+  int rv;
+#ifdef _WIN32
+  // TODO(mehrdadn): This is dangerous on Windows.
+  // We need to keep the actual process handle alive for the PID to stay valid.
+  // Make this change as soon as possible, or the PID may refer to the wrong process.
+  intptr_t handle =
+    spawnvp(P_NOWAIT, worker_command_args_str[0], worker_command_args_str.data());
+  if (handle != -1) {
+    pid = static_cast<int>(GetProcessId((HANDLE)handle));
+    CloseHandle((HANDLE)handle);
+  } else {
+    pid = 0;
   }
+  rv = pid ? 0 : static_cast<int>(GetLastError());
+#else
+  // Launch the process to create the worker.
+  pid = fork();
 
-  // Child process case.
-  // Reset the SIGCHLD handler for the worker.
-  signal(SIGCHLD, SIG_DFL);
+  if (pid == 0) {
+    // Child process case.
+    // Reset the SIGCHLD handler for the worker.
+    // TODO(mehrdadn): Move any work here to the child process itself
+    //                 so that it can also be implemented on Windows.
+    signal(SIGCHLD, SIG_DFL);
 
-  // Try to execute the worker command.
-  std::vector<const char *> worker_command_args_str;
-  for (const auto &arg : worker_command_args) {
-    worker_command_args_str.push_back(arg.c_str());
+    // Try to execute the worker command.
+    std::vector<const char *> worker_command_args_str;
+    for (const auto &arg : worker_command_args) {
+      worker_command_args_str.push_back(arg.c_str());
+    }
+    worker_command_args_str.push_back(nullptr);
+    rv = execvp(worker_command_args_str[0],
+                const_cast<char *const *>(worker_command_args_str.data()));
+  } else {
+    rv = pid;
+  }
+#endif
+  if (rv != 0) {
```

Comment:
Was this a bug in the original code and we always printed the `RAY_LOG(FATAL)` error? That's surprising, I don't recall having seen this error :|

Reply:
No haha, it's because `execvp` doesn't return if it succeeds; it replaces the current process :)

Reply:
Ahahaha I forgot about that corner case :D